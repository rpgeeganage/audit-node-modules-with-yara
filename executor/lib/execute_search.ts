import { exec } from 'child_process';

import debug from 'debug';

const debugExecteSearch = debug('common');

/**
 * Execute the search operation
 *
 * @export
 * @param {string} folderToAudit
 * @param {string} yaraRulesFolder
 * @return {*}  {Promise<string[]>}
 */
export async function executeSearch(
  folderToAudit: string,
  yaraRulesFolder: string
): Promise<string[]> {
  const command = `yara --recursive --print-strings --fail-on-warnings \`find ${yaraRulesFolder} -type f -name "*.yara"\` ${folderToAudit}`;

  const results = await executeShellCommand(command);

  return results
    .split(/\n/)
    .map((s) => s?.trim())
    .filter((s) => s.length);
}

/**
 * Execute shell command
 *
 * @param {string} command
 * @return {*}  {Promise<string>}
 */
export function executeShellCommand(command: string): Promise<string> {
  return new Promise((resolve, reject) => {
    debugExecteSearch('executeShellCommand: [%s]', command);

    exec(command, (error: Error | null, stdOut: string, stdErr: string) => {
      if (error) {
        debugExecteSearch('executeShellCommand: Error occured %o', error);

        return reject(error);
      }

      if (stdErr) {
        debugExecteSearch('executeShellCommand: Std Error occured %s', error);

        return reject(new Error(stdErr));
      }

      debugExecteSearch('executeShellCommand: stdOut %s', stdOut);

      return resolve(stdOut);
    });
  });
}
