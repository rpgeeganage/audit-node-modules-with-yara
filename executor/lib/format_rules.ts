/**
 * Single YARA rule which applied
 *
 * @export
 * @interface ISingleYARARule
 */
export interface ISingleYARARule {
  rule: string;
  string_information: string[];
  [key: string]: string | string[];
}

/**
 * Extract Yara Rules Mapping
 *
 * @param {string} yaraOutput
 * @param {ISavedFile} executableFilePath
 * @return {*}  {ISingleYARARule[]}
 */
export function formatYaraResults(
  outputLines: string[],
  executableFilePath: string
): ISingleYARARule[] {
  const pathRegEx = new RegExp(executableFilePath);
  const ruleIndexes = outputLines.reduce(
    (acc: number[], outline: string, index: number) => {
      if (pathRegEx.test(outline)) {
        acc.push(index);
      }

      return acc;
    },
    []
  );

  return buildRules(ruleIndexes, outputLines);
}

/**
 * Build YARA rule
 *
 * @param {number[]} ruleIndexes
 * @param {string[]} output
 * @return {*}
 */
function buildRules(ruleIndexes: number[], output: string[]) {
  return ruleIndexes.map((v, i: number) => {
    return {
      rule: output[v].split(/\s/, 2)[0],
      string_information: Number.isNaN(ruleIndexes[i + 1] - 1)
        ? output.slice(v + 1)
        : output.slice(v + 1, ruleIndexes[i + 1])
    };
  });
}
