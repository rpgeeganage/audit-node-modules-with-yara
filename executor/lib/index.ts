export * from './execute_search';
export * from './format_rules';
