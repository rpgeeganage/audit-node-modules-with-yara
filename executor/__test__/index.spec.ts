import * as path from 'path';

import * as lib from '../lib';

it('Should run the audit operation', async () => {
  const folderToAudit = path.join(__dirname, 'fixtures/node_module_folder');
  const yaraRuleFolder = path.join(__dirname, 'fixtures/yara_rules');

  const searchResults = await lib.executeSearch(folderToAudit, yaraRuleFolder);
  const formatedResults = lib.formatYaraResults(searchResults, folderToAudit);

  expect(JSON.stringify(formatedResults, null, ' ')).toMatchSnapshot();
});
