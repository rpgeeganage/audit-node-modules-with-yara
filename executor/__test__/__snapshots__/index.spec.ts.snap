// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should run the audit operation 1`] = `
"[
 {
  \\"rule\\": \\"babelcli\\",
  \\"string_information\\": [
   \\"0x6:$name: \\\\\\"name\\\\\\": \\\\\\"babelcli\\\\\\",\\",
   \\"0x1e:$version: \\\\\\"version\\\\\\": \\\\\\"1.0.1\\\\\\"\\"
  ]
 },
 {
  \\"rule\\": \\"cross_env_js\\",
  \\"string_information\\": [
   \\"0x6:$name: \\\\\\"name\\\\\\": \\\\\\"cross-env.js\\\\\\",\\",
   \\"0x22:$version: \\\\\\"version\\\\\\": \\\\\\"5.0.1\\\\\\"\\"
  ]
 }
]"
`;
